src/cobra/io/schema_v2.json

```diff
@@ -0,0 +1,370 @@
+  "title": "COBRA",
+  "description": "JSON representation of COBRA model",
+  "definitions": {
+    "annotation": {
+      "type": "object",
+      "properties": {
+        "sbo": {
+          "type": "string"
+        },
+        "cvterms": {
+          "type": "object",
+          "additionalProperties": {
+            "type": "array",
+            "items": {
+              "type": "object",
+              "properties": {
+                "resources": {
+                  "type": "array",
+                  "items": {
+                    "type": "string"
+                  }
+                },
+                "nested_data": {
+                  "$ref": "#/definitions/annotation"
+                }
+              }
+            }
+          }
+        },
+        "history": {
+          "type": "object",
+          "properties": {
+            "creators": {
+              "type": "array",
+              "items": {
+                "type": "object",
+                "properties": {
+                  "email": {
+                    "type": "string"
+                  },
+                  "given_name": {
+                    "type": "string"
+                  },
+                  "family_name": {
+                    "type": "string"
+                  },
+                  "organisation": {
+                    "type": "string"
+                  }
+                },
+                "required": [
+                  "given_name",
+                  "family_name"
+                ],
+                "additionalProperties": false
+              }
+            },
+            "created_date": {
+              "type": "string",
+              "format": "date-time"
+            },
+            "modified_dates": {
+              "type": "array",
+              "items": {
+                "type": "string",
+                "format": "date-time"
+              }
+            }
+          },
+          "required": [
+            "creators",
+            "created_date",
+            "modified_dates"
+          ],
+          "additionalProperties": false
+        },
+        "listofkeyvalue": {
+          "type": "array",
+          "items": {
+            "type": "object",
+            "properties": {
+              "id": {
+                "type": "string",
+                "pattern": "^[a-zA-Z|_][a-zA-Z0-9_]*$"
+              },
+              "name": {
+                "type": "string"
+              },
+              "key": {
+                "type": "string"
+              },
+              "value": {
+                "type": "string"
+              },
+              "uri": {
+                "type": "string"
+              }
+            },
+            "required": [
+              "key"
+            ],
+            "additionalProperties": false
+          }
+        }
+      }
+    }
+  },
+  "type": "object",
+  "properties": {
+    "id": {
+      "type": "string",
+      "pattern": "^[a-zA-Z|_][a-zA-Z0-9_]*$"
```

Comment:
I think it was. It might be worth having a discussion on JSON standards. 
The IDs for JSON and YAML changed because this patch matches the IDs to SBML.
I can
1) Remove all ID changes
2) Keep them as an option, but have default be no changes
3) Keep them as default.

@matthiaskoenig - you seem to prefer no changes in SBML if I recall correctly. Do you have an opinion on replacement in JSON?

Reply:
I would prefer no id changes at all. Id changes just create so many issues. I highly would recommend to get rid of all id changes besides:
- JSON import (IDs are fixed to be valid SIds)
- MAT import (IDS are fixed to be valid SIds)

It would be great if we could have a vote/decision on this because most of our issues are related to the nonsense id replacements. There are hundreds of lines of code to deal with this nonsense which create more and more issues and still don't catch all corner cases.

Reply:
What do you mean by SIds?
SBML IDs?
If JSON IDs are fixed to be valid SBML IDs, there must be an id change, because JSON allows IDs that are not SBML compliant. Also, cobrapy allows IDs that are not SBML compliant. If there is a change when reading JSON files, then there are usually changes when writing JSON files, and we're back at the "nonsense id replacement."

We could decide that there are no ID changes when reading JSON/MAT/YAML (as long as the ID is valid for cobra), and then make each format responsible for converting IDs to its own format.
So JSON can have an id that looks like "3dpg", which is not valid SBML ID, and if I read that model into cobra, and then write it as SBML, it will be transformed to meta id "M_3dpg" when writing SBML.

So each format does
File -> Cobrapy format when reading
cobrapy format -> file limitations when writing

Reply:
Yes, SIds are valid SBML ids. This can be checked very easily with a regex.

Like mentioned above the only id changes are on
- JSON import (IDs are fixed to be valid SIds)
- MAT import (IDS are fixed to be valid SIds)
- OTHER_FORMAT_NOT_SBML (IDS are fixed to be valid SIds)

There would be a single function fixing ids, e.g. `fix_invalid_ids` used in both/all importers to have a consistent behavior across formats. Nothing would be replaced fixed on any exports, because SIds are valid JSON and MAT Ids.

Reply:
This would still not affect models build directly with cobrapy objects. And if we substitute IDs there on object creation it would be hard to get backwards compatibility for older code, especially with the JSON v1 schema which is fixed/won't be changed anymore.

However, I see your point for JSON v2 (this discussion) and I think it makes sense to have it as a "fully SBML-compliant without the XML" format. We can maybe just provide a helper function to fix IDs and direct users to that if they want to write it and the IDs are not compliant.